### PR TITLE
Much reduced memory usage of collision values in importer

### PR DESCRIPTION
**Pull request by tinwelint**
- Reduced big tracker cache from 6 --> 5 bytes, i.e. an overall ~8% lowered memory usage for large imports
- Moved keeping of collision values off-heap and reduced down to 1/6th of the memory it used to occupy in heap
- Removed array for keeping duplicate node ids entirely

These changes are targeted to improve scenarios where there are lots of duplicate node IDs in the input files. Previously the management of those duplicates required a lot of heap memory. Now this has changed to require only a fraction of that, and also moved off-heap.

There is now only remnants of the page cache meta data left in the heap that needs to be considered when chosing heap size for an import, nothing else by design requires a heap larger than a couple of gigabytes, tops, for an import regardless of data size and its condition.

### CODE REVIEW

**Comment by tinwelint** [comment]
Builds on top of #10648

File: community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/EncodingIdMapper.java [hunk 1]
```diff
@@ -678,6 +732,7 @@ private long dataValue( long index )
         return clearCollision( dataCache.get( trackerCache.get( index ) ) );
     }
 
+//<<<<<<< HEAD
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/EncodingIdMapper.java]: Merge left-overs.

File: community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/StringCollisionValues.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            throw new IllegalArgumentException( string );
+        }
+
+        long endOffset = offset + 2 + bytes.length;
+        ByteArray end = cache.at( endOffset );
+        if ( end != current )
+        {
+            offset = endOffset;
+            current = end;
+        }
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/StringCollisionValues.java]: Are we sure that the value will always fit within a single chunk?

File: community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/StringCollisionValues.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+        long startOffset = offset;
+        current.setByte( offset++, 0, (byte) (length & 0xFF) );
+        current.setByte( offset++, 0, (byte) ((length >>> 8) & 0xFF) );
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/StringCollisionValues.java]: Why not getShort and setShort?

File: community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/StringCollisionValuesTest.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        // given
+        try ( StringCollisionValues values = new StringCollisionValues( factory, 100_000 ); )
+        {
+            // when
+            long[] offsets = new long[100];
+            String[] strings = new String[offsets.length];
+            for ( int i = 0; i < offsets.length; i++ )
+            {
+                String string = random.string();
```
Comment (chrisvest) [inline community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/StringCollisionValuesTest.java:76]: By default, the maximum length of string that this will produce is 20 characters. I think that misses a lot of important edge cases.

**Review by chrisvest — COMMENTED** [review]
Looks okay. Just a few comments. Please take a look.

File: community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/StringCollisionValues.java (hunk 3 above)
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/StringCollisionValues.java]: This is because we can cross a dynamic byte array boundary underneath. It's a different use of ByteArray, one where it's truly used as a byte[]. Doing it this way will work around issues with this use case.

Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/StringCollisionValues.java]: Oh wait, we work-around this right above this. I think this a remnant from when that wasn't done. I'll change it.

Comment (chrisvest) [reply community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/StringCollisionValues.java]: Yeah, it's a bit hard to follow when you need to be weary of the boundary or not.

**Comment by tinwelint** [comment]
Added testing for larger keys, much larger and your concerns came true. Updated the implementation to handle all that.

**Review by chrisvest — APPROVED** [review]
